File: backend/app/services/projects.py
from ..schemas.contracts import ProjectInformation, ProjectInput, ProjectStage, RiskPrediction
from .prediction import js_round, risk_level

import json
from pathlib import Path

SEED_PATH = Path(__file__).resolve().parents[2] / "data" / "projects.json"


def bundled_projects() -> list[ProjectInput]:
    seeds = json.loads(SEED_PATH.read_text())["seeds"]
    projects = []
    for index, row in enumerate(seeds):
        name, state, district, kind, total, acquired, paid, cases, days, complexity, latitude, longitude = row
        projects.append(ProjectInput(
            id=f"LS-2026-{index + 1:03d}", name=name, state=state, district=district, type=kind,
            total_parcels=total, acquired_parcels=acquired, compensation_paid=paid,
            legal_cases=cases, approval_days=days, complexity=complexity,
            latitude=latitude, longitude=longitude,
            landowners=int(js_round(total * (1 + complexity * 0.12))),
            compensation_budget_cr=js_round(total * 0.38),
            approvals_pending=3 if days > 80 else 2 if days > 30 else 1 if days > 0 else 0,
            clearance_status="Clearances delayed" if days > 60 else "Under review",
            expected_completion="2026-12-31" if index < 8 else "2027-03-31" if index < 16 else "2027-06-30",
            agency=f"{state} Project Coordination Unit (demo)",
        ))
    return projects


class DemoProjectRepository:
    def list_projects(self) -> list[ProjectInput]:
        return bundled_projects()

    def get_project(self, project_id: str) -> ProjectInput | None:
        return next((p for p in self.list_projects() if p.id == project_id), None)


def project_information(project: ProjectInput, prediction: RiskPrediction) -> ProjectInformation:
    progress = js_round(project.acquired_parcels / project.total_parcels * 100)
    rows = [
        ("Land identification", 100, 0),
        ("Notification", min(100, progress + 45), 5),
        ("Valuation", min(100, project.compensation_paid + 20), 12),
        ("Compensation", project.compensation_paid, int(js_round((100 - project.compensation_paid) * 0.6))),
        ("Possession", progress, int(js_round((100 - progress) * 0.4))),
        ("Legal resolution", max(0, 100 - project.legal_cases * 4), project.legal_cases * 2),
        ("Final acquisition", progress, int(js_round(project.approval_days * 0.3))),
    ]
    stages = [
        ProjectStage(
            name=name,
            completion=completion,
            delay_days=0 if completion == 100 else delay,
            risk=risk_level(10 if completion == 100 else min(100, 100 - completion + delay / 2)),
            status="Complete" if completion == 100 else "Not started" if completion == 0 else "Needs attention" if completion < 45 else "In progress",
        )
        for name, completion, delay in rows
    ]
    primary_risk = max(prediction.factors, key=lambda factor: factor.contribution).name if prediction.factors else "Per-project ML explanation unavailable"
    if prediction.explanation is not None:
        explanation = prediction.explanation
        primary_risk = (explanation.top_risk_factors[0].name if explanation.top_risk_factors else "No risk-increasing factors relative to the model baseline") if explanation.status == "available" else "Per-project ML explanation unavailable"
    values = project.model_dump()
    values.update(
        pending_parcels=project.total_parcels - project.acquired_parcels,
        progress=progress,
        risk_score=prediction.score,
        expected_delay=prediction.delay_days,
        risk_level=prediction.level,
        risk_change=8 if prediction.score > 80 else 4 if prediction.score > 60 else -5,
        primary_risk=primary_risk,
        status="Delayed" if prediction.score > 80 else "At risk" if prediction.score > 60 else "On track",
        prediction=prediction,
        stages=stages,
    )
    return ProjectInformation.model_validate(values)

File: lib/landsight/projects.ts
import { PROJECTS, PROJECT_TYPES } from './data'
import { validateMLPrediction } from './ml'
import type { Project } from './types'

export interface ProjectCollection {
  projects: Project[]
  source: 'postgresql' | 'demo-fallback'
  notice: string
  predictionsPersisted: boolean
}

const levels = ['LOW', 'MEDIUM', 'HIGH', 'CRITICAL']
const fallbackNotice = 'Demo fallback — backend or database unavailable. Showing bundled synthetic projects.'

function invalid(): never {
  throw new Error('The project API returned invalid data. Retry after checking the backend; no demo data was substituted.')
}

export function validateProject(value: unknown): Project {
  if (!value || typeof value !== 'object') return invalid()
  const p = value as Project
  for (const key of ['id', 'name', 'state', 'district', 'clearanceStatus', 'expectedCompletion', 'agency', 'primaryRisk'] as const) {
    if (typeof p[key] !== 'string' || !p[key].trim()) return invalid()
  }
  for (const key of ['totalParcels', 'acquiredParcels', 'pendingParcels', 'landowners', 'legalCases', 'approvalDays', 'approvalsPending', 'complexity', 'riskScore', 'expectedDelay'] as const) {
    if (!Number.isSafeInteger(p[key]) || p[key] < 0) return invalid()
  }
  for (const key of ['compensationPaid', 'compensationBudgetCr', 'progress', 'latitude', 'longitude', 'riskChange'] as const) {
    if (!Number.isFinite(p[key])) return invalid()
  }
  if (!/^[A-Za-z0-9_-]{1,128}$/.test(p.id) || !PROJECT_TYPES.includes(p.type) || p.source !== 'Synthetic'
    || !levels.includes(p.riskLevel) || !['On track', 'At risk', 'Delayed'].includes(p.status)
    || !Number.isFinite(Date.parse(p.expectedCompletion)) || p.totalParcels <= 0 || p.acquiredParcels > p.totalParcels
    || p.pendingParcels !== p.totalParcels - p.acquiredParcels || p.compensationPaid < 0 || p.compensationPaid > 100
    || p.compensationBudgetCr < 0 || p.complexity < 1 || p.complexity > 5 || p.approvalDays > 180
    || Math.abs(p.latitude) > 90 || Math.abs(p.longitude) > 180 || p.riskScore > 100
    || Math.abs(p.progress - p.acquiredParcels / p.totalParcels * 100) > 0.500001) return invalid()
  const prediction = p.prediction
  if (!prediction || prediction.score !== p.riskScore || prediction.level !== p.riskLevel
    || prediction.delayDays !== p.expectedDelay || prediction.isDemo !== true || !Array.isArray(prediction.factors)
    || typeof prediction.model !== 'string' || (prediction.metadata?.status !== undefined && !['trained', 'demo'].includes(prediction.metadata.status))) return invalid()
  if (prediction.metadata?.status === 'trained') validateMLPrediction(prediction)
  if (!prediction.factors.every(f => f && typeof f.id === 'string' && typeof f.name === 'string'
    && typeof f.description === 'string' && Number.isFinite(f.contribution))) return invalid()
  if (!Array.isArray(p.stages) || p.stages.length !== 7 || !p.stages.every(s => s && typeof s.name === 'string'
    && Number.isFinite(s.completion) && s.completion >= 0 && s.completion <= 100 && levels.includes(s.risk)
    && Number.isInteger(s.delayDays) && s.delayDays >= 0
    && ['Complete', 'In progress', 'Needs attention', 'Not started'].includes(s.status))) return invalid()
  return p
}

async function requestProjects(path: string): Promise<Response | null> {
  let response: Response
  try {
    response = await fetch(path, { cache: 'no-store', signal: AbortSignal.timeout(25_000) })
  } catch {
    return null
  }
  if (response.ok || response.status === 404) return response
  const error = await response.json().catch(() => null)
  if (response.status === 503 && error?.error?.code === 'BACKEND_UNAVAILABLE') return null
  throw new Error(error?.error?.code === 'PROJECT_SOURCE_UNAVAILABLE'
    ? 'The project database is unavailable and backend demo fallback is disabled.'
    : 'The project API could not load this view. No demo data was substituted for a backend error.')
}

export async function fetchProjectCollection(): Promise<ProjectCollection> {
  const response = await requestProjects('/api/projects')
  if (!response) return { projects: PROJECTS, source: 'demo-fallback', notice: fallbackNotice, predictionsPersisted: false }
  if (!response.ok) return invalid()
  const source = response.headers.get('X-LandSight-Data-Source')
  if (source !== 'postgresql' && source !== 'demo-fallback') return invalid()
  const value: unknown = await response.json().catch(() => invalid())
  if (!Array.isArray(value)) return invalid()
  const projects = value.map(validateProject)
  if (new Set(projects.map(p => p.id)).size !== projects.length) return invalid()
  return {
    projects, source, notice: source === 'postgresql' ? 'PostgreSQL / PostGIS — persisted synthetic data, not government records.' : fallbackNotice,
    predictionsPersisted: response.headers.get('X-LandSight-Predictions-Persisted') === 'true',
  }
}

export async function fetchProject(id: string): Promise<Project | undefined> {
  if (!/^[A-Za-z0-9_-]{1,128}$/.test(id)) return undefined
  const response = await requestProjects(`/api/projects/${encodeURIComponent(id)}`)
  if (!response) return PROJECTS.find(p => p.id === id)
  if (response.status === 404) return undefined
  const project = validateProject(await response.json().catch(() => invalid()))
  if (project.id !== id) return invalid()
  return project
}
